lib/player.js

```diff
@@ -5691,6 +5701,49 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
     }
   }
 
+  /**
+   * Callback for liveSync
+   *
+   * @private
+   */
+  onTimeUpdate_() {
+    // If the live stream has reached its end, do not sync.
+    if (!this.isLive()) {
+      return;
+    }
+    const seekRange = this.seekRange();
+    if (!Number.isFinite(seekRange.end)) {
+      return;
+    }
+    const currentTime = this.video_.currentTime;
+    if (currentTime > seekRange.start) {
+      // Bad stream?
+      return;
+    }
+    const liveSyncMaxLatency = this.config_.streaming.liveSyncMaxLatency;
+    const liveSyncPlaybackRate = this.config_.streaming.liveSyncPlaybackRate;
+    const playbackRate = this.video_.playbackRate;
+    const latency = seekRange.end - this.video_.currentTime;
+    let offset = 0;
+    // HLS in SRC= doesn't update every time the seekrange, so we need calculate
+    // a safe offset
+    if (this.loadMode_ == shaka.Player.LoadMode.SRC_EQUALS) {
+      const buffered = this.video_.buffered;
+      if (buffered.length > 0) {
+        const bufferedEnd = buffered.end(buffered.length - 1);
+        offset = Math.max(liveSyncPlaybackRate, bufferedEnd - seekRange.end);
```

Comment:
Another option is put the min in 0, but I'd prefer the current option

Reply:
Your explanation makes sense, but I don't find the code clear enough on its own without this conversation.  And I will definitely forget.

Here's a comment that I think might help, to replace the one above:

```js
// In src= mode, the seek range isn't updated frequently enough, so we need to fudge
// the latency number with an offset.  The playback rate is used as an offset, since that
// is the amount we catch up 1 second of accelerated playback.
```

I still don't understand the other duration, though.  Why do you subtract bufferedEnd - seekRange.end?  That should be negative if we've fallen behind, right?

```
-------|---------------|-------------|--------------|------
seekRange.start   currentTime   bufferedEnd   seekRange.end
```

Reply:
bufferedEnd can greater than seekRange.end when the current time is near to live edge. 

I’ll add your comment tomorrow. Thanks!

Reply:
This Math.max involving the bufferedEnd - seekRange.end parameter still doesn't make sense to me.

Here's a concrete scenario.  Let's say your latency goal is 3s.  You want to be within 3s of the end of the seek range.

currentTime = 10
seekRange.end = 15
bufferedEnd = 18
liveSyncPlaybackRate = 1.1

latency = seekRange.end - currentTime = 5

offset = Math.max(liveSyncPlaybackRate, bufferedEnd - seekRange.end)
offset = Math.max(1.1, 18 - 15)
offset = 3

latency - offset = 2, which is less than our target of 3.  So even though we're 5s behind the end of the seek range, we don't accelerate playback.  And because we've buffered past the end of the seek range, we _could_ accelerate playback _very_ safely.

So it seems to me that in the cases where bufferedEnd > seekRange.end, the offset value has the opposite effect of what it should do.

Am I crazy?

Reply:
You're right, but if we take into account Theodore's comment, the event should trigger at most every 250ms, so it would still be fine. (https://developer.mozilla.org/en-US/docs/Web/API/HTMMLediaElement/timeupdate_event)

Note: this is for valid, because the src= for live should only be used in Safari.

Reply:
4Hz is the minimum frequency, not the maximum. The maximum is 66Hz, which would be one call every 15ms.

Reply:
In the case of the maximum, we have no problem :)